### DISCUSSION TITLE
Any suggestions on how to troubleshoot: "mroute: pending queue full, dropping entries"

### DISCUSSION THREAD

**Discussion by henk52** [troglobit/mrouted#49] [Q&A] [answered]
Hi

I was wondering if anyone had any suggestions on how to troubleshoot: "mroute: pending queue full, dropping entries"

It seems, to me, to be related to the kernel asking mrouted for cache information.

OS: Ubuntu 18.04.5
mrouted: 3.9.8, 4.1 and 4.2
mrouted is running without a config file.
The binaries where build from the tar.gz.
The 3.9.8 and 4.1 was built on an Ubuntu 18.04
The 4.2 was built on an Ubuntu 20.04
RAM: 16GB

I've run the mrouted daemons with debug=cache, but I can't see any output that I can correlate to when the errors occur.

Going to a  VM with 64GB seemed to reduce how often the error is reported.

The error seems to come from the kernel, source file: /linux/net/ipv4/ipmr.c
I'm guessing the the kernel asks mrouted for a cachereport, possibly via netlink, I'm guessing, via the socket, in the code in sk_filter_trim_cap() in the /linux/net/core/filter.c  file.

1. Could this be related to the mrouted?
2. Any idea of what the problem might be?
3. Any idea of how I might debug mrouted, to identify if mrouted is running out of resources?
 
  Thanks

**Comment by troglobit** [comment, marked as answer]
Sounds like you have a LOT of multicast coming in to your router. It's possible you see this because mrouted is not quick enough to process the IGMPMSG_NOCACHE messages from the kernel.  This message is sent on the mutlicast routing socket for any new unknown multicast (S,G) pair.  If the routing daemon knows of a receiver on another LAN that want this it creates an mroute for the corresponding outbound interface, if *not* then it takes note of the (S,G) pair and creates a "block" rule to prevent further notifications from the kernel.  Such block rules have no outbound interfaces.

The function responsible for creating all mroutes is `k_add_rg()`, called from `prune.c:add_table_entry()`.  You can add some more logs there to see if you hit those sections.

**Reply by henk52** [reply]
Thank you very much for your reply, that is very helpful.

**Reply by troglobit** [reply]
No problem. You can press the mark at the top-right of my reply to set the Accepted Answer.  Similar to Stackoverflow, I guess.
